proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,302 @@
+to restart if validators in the restart can reach agreement, or print debug
+information and halt otherwise.
+
+## New Terminology
+
+* "cluster restart": When there is an outage such that the whole cluster
+stalls, human may need to restart most of the validators with a sane state so
+that the cluster can continue to function. This is different from sporadic
+single validator restart which does not impact the cluster. See
+[cluster restart](https://docs.solana.com/running-validator/restart-cluster)
+for details.
+
+* "optimistically confirmed block": a block which gets the votes from the
+majority of the validators in a cluster (> 2/3 stake). Our algorithm tries to
+guarantee that an optimistically confirmed will never be rolled back. When we 
+are performing cluster restart, we normally start from the highest 
+optimistically confirmed block, but it's also okay to start from a child of the 
+highest optimistically confirmed block as long as consensus can be reached.
+
+* "silent repair phase": In the new repair and restart plan, the validators in
+restart will first spend some time to exchange information, repair missing
+blocks, and finally reach consensus. The validators only continue normal block
+production and voting after consensus is reached. We call this preparation
+phase where block production and voting are paused the silent repair phase.
+
+* "ephemeral shred version": right now we update `shred_version` during a
+cluster restart, it is used to verify received shreds and filter Gossip peers. 
+In the new repair and restart plan, we introduce a new temporary shred version 
+in the silent repair phase so validators in restart don't interfere with those 
+not in restart. Currently this ephemeral shred version is calculated using
+`(current_shred_version + 1) % 0xffff`.
+
+* `RESTART_STAKE_THRESHOLD`: We need enough validators to participate in a
+restart so they can make decision for the whole cluster. If everything works
+perfect, we only need 2/3 of the total stake. However, validators could die
+or perform abnormally, so we currently set the `RESTART_STAKE_THRESHOLD` at
+80%, which is the same as now.
+
+## Motivation
+
+Currently during a cluster restart, validator operators need to decide the
+highest optimistically confirmed slot, then restart the validators with new
+command-line arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it is
+detrimental to the viability of the ecosystem.
+
+We aim to automate the negotiation of highest optimistically confirmed slot and
+the distribution of all blocks on that fork, so that we can lower the 
+possibility of human mistakes in the cluster restart process. This also reduces 
+the burden on validator operators, because they don't have to stay around while 
+the validators automatically try to reach consensus, they will be paged if 
+things go wrong.
```

Comment:
"they will be paged if things go wrong." -- this feels slightly like editorializing as we don't actually add a paging facility on this proposal, although we may want to suggest under what condition a paging solution could be triggered (unexpected validator exit?)

Reply:
Changed, does this look better?